compiler/src/iree/compiler/GlobalOptimization/FuseDequantizationMatmul.cpp

```diff
@@ -0,0 +1,1010 @@
+  if (!elementTypeOut.isIntOrFloat()) {
+    return failure();
+  }
+  unsigned bitWidthOut = elementTypeOut.getIntOrFloatBitWidth();
+  auto elementTypeIn =
+      llvm::cast<ShapedType>(genericOp.getInputs()[0].getType())
+          .getElementType();
+  if (!elementTypeIn.isIntOrFloat()) {
+    return failure();
+  }
+  unsigned bitWidthIn = elementTypeIn.getIntOrFloatBitWidth();
+  if (bitWidthIn >= bitWidthOut) {
+    return failure();
+  }
+
+  return success();
+}
+
+// Creates a new flow.dipatch.region op and places the
+// passed ops inside as long as the dequant op is a
+// producer for the matmul op
+static LogicalResult fuseDequantAndMatmul(RewriterBase &rewriter,
+                                          Operation *dequant, Operation *matmul,
+                                          std::optional<Operation *> fill) {
+
+  auto regionOp = matmul->getParentOfType<IREE::Flow::DispatchRegionOp>();
+  if (!regionOp) {
+    FailureOr<IREE::Flow::DispatchRegionOp> maybeRegionOp =
+        IREE::Flow::wrapOpInDispatchRegion(rewriter, matmul);
+    if (failed(maybeRegionOp))
+      return failure();
+    regionOp = maybeRegionOp.value();
+  }
+
+  FailureOr<IREE::Flow::DispatchRegionOp> maybeFusedRegionOp =
+      IREE::Flow::clonePrecedingOpIntoDispatchRegion(rewriter, dequant,
+                                                     regionOp);
+  if (failed(maybeFusedRegionOp))
+    return failure();
+
+  if (fill && *fill) {
+    FailureOr<IREE::Flow::DispatchRegionOp> maybeFusedFillRegionOp =
+        IREE::Flow::clonePrecedingOpIntoDispatchRegion(rewriter, fill.value(),
+                                                       regionOp);
+    if (failed(maybeFusedFillRegionOp))
+      return failure();
+  }
+
+  return success();
+}
+
+static FailureOr<IREE::Flow::DispatchRegionOp>
+wrapConsecutiveOpsInDispatchRegion(RewriterBase &rewriter,
+                                   SmallVector<Operation *> ops) {
+  FailureOr<IREE::Flow::DispatchRegionOp> maybeRegionOp =
+      IREE::Flow::wrapOpInDispatchRegion(rewriter, ops.back());
+  if (failed(maybeRegionOp)) {
+    return failure();
+  }
+  IREE::Flow::DispatchRegionOp regionOp = maybeRegionOp.value();
+
+  SmallVector<Operation *> precedingOps(ops.begin(), ops.end() - 1);
+  FailureOr<IREE::Flow::DispatchRegionOp> maybeFusedRegionOp =
+      IREE::Flow::movePrecedingOpsIntoDispatchRegion(rewriter, precedingOps,
+                                                     regionOp);
+  if (failed(maybeFusedRegionOp)) {
+    return failure();
+  }
+  regionOp = maybeFusedRegionOp.value();
+
+  return regionOp;
+}
+
+static SmallVector<utils::IteratorType>
+getParallelAndReductionIterators(unsigned nLoops, unsigned nReduction) {
+  SmallVector<utils::IteratorType> res(nLoops - nReduction,
+                                       utils::IteratorType::parallel);
+  res.append(nReduction, utils::IteratorType::reduction);
+  return res;
+}
+
+struct QuantizedMatmulRewriter {
```

Comment:
Total style nit (i.e. no need to implement) : I prefer just individual `static` methods, which are just a different way to carry state.. this is fine as well.